core/src/com/cloud/agent/resource/virtualnetwork/facade/QuaggaConfigItem.java

```diff
@@ -0,0 +1,93 @@
+import com.cloud.utils.net.cidr.CIDR;
+
+public class QuaggaConfigItem extends AbstractConfigItemFacade {
+
+    @Override
+    public List<ConfigItem> generateConfig(final NetworkElementCommand cmd) {
+        final QuaggaConfigCommand command = (QuaggaConfigCommand)cmd;
+
+        final String vrIp = command.getRouterPublicIp();
+        final String vrName = command.getRouterName();
+        final CIDR[] cidrs = command.getTierCidrs();
+        final OSPFZoneConfig qzc = command.getZoneConfig();
+
+        StringBuilder zebra = new StringBuilder();
+        zebra.append("hostname " + vrName).append(",");
+        zebra.append("interface eth1").append(",");
+        zebra.append(" description link to area 0").append(",");
+        zebra.append("  ip address " + vrIp + "/24").append(",");
+        zebra.append("  link-detect").append(",");
+        zebra.append("log file /var/log/quagga/zebra.log").append(",");
+
+        StringBuilder ospfd = new StringBuilder();
+        ospfd.append("hostname " + vrName).append(",");
+        ospfd.append("interface eth1").append(",");
+        ospfd.append("  ip ospf hello-interval ").append(qzc.getHelloInterval()).append(",");
+        ospfd.append("  ip ospf dead-interval ").append(qzc.getDeadInterval()).append(",");
+        ospfd.append("  ip ospf retransmit-interval ").append(qzc.getRetransmitInterval()).append(",");
+        ospfd.append("  ip ospf transmit-delay ").append(qzc.getTransitDelay()).append(",");
+        if (qzc.getAuthentication().equals(OSPFZoneConfig.Authentication.MD5) && qzc.getPassword().length() > 3){
+            ospfd.append("  ip ospf authentication message-digest").append(",");
+            ospfd.append("  ip ospf message-digest-key 1 md5 ").append(qzc.getPassword()).append(",");
+        }
+        ospfd.append("router ospf").append(",");
+        ospfd.append("  ospf router-id " + vrIp).append(",");
+        ospfd.append("  redistribute connected").append(",");
+        ospfd.append("  no passive-interface eth1").append(",");
+        ospfd.append("  network " + vrIp + "/24 area ").append(qzc.getOspfArea()).append(",");
+        if (qzc.getAuthentication().equals(OSPFZoneConfig.Authentication.MD5) && qzc.getPassword().length() > 3){
+            ospfd.append("  area ").append(qzc.getOspfArea()).append(" authentication message-digest").append(",");
+        }
+        if (cidrs != null) {
+            for (CIDR cidr : cidrs) {
+                ospfd.append("  network " + cidr + " area ").append(qzc.getOspfArea()).append(",");
+            }
+        }
+        ospfd.append("log file /var/log/quagga/ospfd.log").append(",");
+
+        final String tmpCfgFilePath = "/etc/quagga/";
+        final String tmpCfgFileName = "ospfd.conf.new." + String.valueOf(System.currentTimeMillis());
+
+        final QuaggaRule quaggaRule = new QuaggaRule(zebra.toString(), ospfd.toString(), tmpCfgFilePath, tmpCfgFileName, vrIp);
+
+        return generateConfigItems(quaggaRule);
+    }
```

Comment:
This method appears to work with the following concerns:

1. A common zerba/ospfd configuration format (lines of space separated values ending with a ``,`` with child items idented with a two (2) spaces).  It also includes conversion of various types to ``String``.
1. Generation of a zebra configuration
1. Generation of an ospfd configuration

The mix of these concerns has created repeating groups and a method that is larger than necessary.  Please consider decomposing as follows:

1. Create a ``QuaggaConfFileBuilder`` class that encapsulates the configuration file format concern.  This class could methods such as ``QuaggaConfFileBuilder withDirective(String... elements)`` that joins the passed ``elements`` with a space and appends a comma; ``QuaggaConfFileBuilder addChildDirective()`` which returns a nested builder to represent the indentation hierarchy; etc.
1. Split the creation of the zebra and ospf configurations into separate private methods or individual small classes.  The decision as to which approach would be most effective would depend on how many rules/data type conversions need to be represented.